### ISSUE TITLE
_is_zipfile() in torch.load is expensive

### ISSUE THREAD

**Issue by meakbiyik** [pytorch/pytorch#108955]
### 🐛 Describe the bug

Hello everyone, I have written a small package called [`torchcache`](https://github.com/meakbiyik/torchcache) to cache and serve the outputs of PyTorch Modules for hefty, pre-trained networks. However, the performance of the serving functionality appeared to be quite random, sometimes becoming outright unusable. After a couple rounds of optimizations, I ran a detailed profiler on the code and noticed that in `torch.load` calls, the most time-consuming operation was in fact `_is_zipfile` check. Here's the output of the relevant section in the profiler, the total training duration was 3189 seconds.

```
Profile stats for: [Strategy]SingleDeviceStrategy.training_step
         2335008 function calls (2309731 primitive calls) in 3165.676 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       10    0.000    0.000 3165.676  316.568 strategy.py:372(training_step)
       10    0.006    0.001 3165.675  316.567 gps_sam_comparison.py:185(training_step)
 7420/350    0.017    0.000 3165.594    9.045 module.py:1514(_wrapped_call_impl)
 7420/350    0.035    0.000 3165.592    9.045 module.py:1520(_call_impl)
       70    0.006    0.000 3163.539   45.193 GVTNet.py:248(preprocess_batch)
       60    0.009    0.000 3163.524   52.725 GVTNet.py:371(_forward_video)
       90    0.004    0.000 3162.161   35.135 torchcache.py:229(forward_pre_hook)
       90    0.899    0.010 3162.089   35.134 torchcache.py:376(_fetch_cached_embeddings)
    11909    0.358    0.000 3160.457    0.265 torchcache.py:571(_load_from_file)
    11909    0.394    0.000 3128.455    0.263 serialization.py:856(load)
    11909    0.203    0.000 2661.872    0.224 serialization.py:109(_is_zipfile)
    47636 2661.602    0.056 2661.602    0.056 {method 'read' of '_io.BufferedReader' objects}
       40    0.003    0.000 1206.576   30.164 GVTNet.py:167(forward)
    11909  296.466    0.025  296.496    0.025 serialization.py:446(__init__)
    11909  126.933    0.011  141.179    0.012 serialization.py:1317(_load)
```

Note that the actual tensor loading, `_load`, only takes 141 seconds in total, while the `_is_zipfile` check takes a whopping 2661 seconds in comparison. `_is_zipfile` is a very simple function, I am copy-pasting it from the master branch below:

```
def _is_zipfile(f) -> bool:
    # This is a stricter implementation than zipfile.is_zipfile().
    # zipfile.is_zipfile() is True if the magic number appears anywhere in the
    # binary. Since we expect the files here to be generated by torch.save or
    # torch.jit.save, it's safe to only check the start bytes and avoid
    # collisions and assume the zip has only 1 file.
    # See bugs.python.org/issue28494.

    # Read the first 4 bytes of the file
    read_bytes = []
    start = f.tell()

    byte = f.read(1)
    while byte != b"":
        read_bytes.append(byte)
        if len(read_bytes) == 4:
            break
        byte = f.read(1)
    f.seek(start)

    local_header_magic_number = [b'P', b'K', b'\x03', b'\x04']
    return read_bytes == local_header_magic_number
```

I imagine the expensive operation here  is the repeated `read` calls, though even then `_load` is several orders of magnitude faster, so I am not sure about the exact culprit. Overall, considering that we can know in advance whether the file is a zipfile or not, a quick fix to this issue would be to allow providing an argument that corresponds to the one in `torch.save`, `_use_new_zipfile_serialization`, and skip this check. Indeed, when I replace this check with a `return True`, the fetching becomes 3x faster.

```
Profile stats for: [Strategy]SingleDeviceStrategy.training_step
         2121584 function calls (2096307 primitive calls) in 1101.175 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
       10    0.000    0.000 1101.175  110.118 strategy.py:372(training_step)
       10    0.006    0.001 1101.174  110.117 gps_sam_comparison.py:185(training_step)
 7420/350    0.016    0.000 1101.084    3.146 module.py:1514(_wrapped_call_impl)
 7420/350    0.036    0.000 1101.083    3.146 module.py:1520(_call_impl)
       70    0.006    0.000 1098.966   15.700 GVTNet.py:248(preprocess_batch)
       60    0.009    0.000 1098.950   18.316 GVTNet.py:371(_forward_video)
       90    0.004    0.000 1097.552   12.195 torchcache.py:229(forward_pre_hook)
       90    0.787    0.009 1097.472   12.194 torchcache.py:376(_fetch_cached_embeddings)
    11840    7.340    0.001 1095.938    0.093 torchcache.py:571(_load_from_file)
    11840    0.383    0.000 1072.399    0.091 serialization.py:857(load)
    11840 1014.314    0.086 1014.344    0.086 serialization.py:447(__init__)
       40    0.003    0.000  424.986   10.625 GVTNet.py:167(forward)
    11840   35.263    0.003   43.085    0.004 serialization.py:1318(_load)
```

It still feels a bit slow tbh, particularly the weirdness that the main time consumer is now the zipfile reader __init__, which might in fact be a bug. Also, despite the latest version having the `mmap` option, the zipfile reader still does `read()` calls to the system to get information instead of reading it from mmapped file directly (see `load` function in `serialization.py`).

<details><summary>Versions</summary>
<p>

PyTorch version: 2.1.0.dev20230830+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.13.4
Libc version: glibc-2.28

Python version: 3.10.3 (main, Jan  3 2023, 08:48:12) [GCC 8.3.0] (64-bit runtime)
Python platform: Linux-4.19.0-25-amd64-x86_64-with-glibc2.28
Is CUDA available: False
CUDA runtime version: 10.1.243
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       36 bits physical, 48 bits virtual
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               58
Model name:          Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz
Stepping:            9
CPU MHz:             1596.676
CPU max MHz:         3900.0000
CPU min MHz:         1600.0000
BogoMIPS:            6984.40
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            8192K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm cpuid_fault epb pti tpr_shadow vnmi flexpriority ept vpid fsgsbase smep erms xsaveopt dtherm ida arat pln pts

Versions of relevant libraries:
[pip3] flake8==6.1.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.2
[pip3] pytorch-lightning==2.0.7
[pip3] pytorch-triton==2.1.0+e6216047b8
[pip3] torch==2.1.0.dev20230830+cu118
[pip3] torchcache==0.3.1
[pip3] torchmetrics==1.0.1
[pip3] torchvision==0.16.0.dev20230830+cu118

</p>
</details> 

cc @mruberry @mikaylagawarecki

**Comment by tringwald** [comment]
It seems like you're `torch.load`ing a file 11909 times from disk, is this really the desired behavior?
```
    ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    [...]
    11909    0.358    0.000 3160.457    0.265 torchcache.py:571(_load_from_file)
    11909    0.394    0.000 3128.455    0.263 serialization.py:856(load)
    11909    0.203    0.000 2661.872    0.224 serialization.py:109(_is_zipfile)
    47636 2661.602    0.056 2661.602    0.056 {method 'read' of '_io.BufferedReader' objects}
```

I also can't reproduce these timings with the given `torch` version or the latest nightly builds. For me, both the `read` call and `PyTorchFileReader`'s init were in the millisecond/sub-millisecond region on my local SSD, while they seem to be in the 56ms/25ms region for you. Are you by any chance reading from an old HDD or a network share?
```
    47636 2661.602    0.056 2661.602    0.056 {method 'read' of '_io.BufferedReader' objects}
    11909  296.466    0.025  296.496    0.025 serialization.py:446(__init__)
```

With regard to mmap'ing, you might want to try passing an already mmap'ed handle to `torch.load` and see if that works out for you.

```python
import mmap
with open('testfile.pth', 'rb+') as f:
    mm = mmap.mmap(f.fileno(), 0)
    torch.load(mm)
```

**Comment by meakbiyik** [comment]
Hi @tringwald - thank you for checking it out! That is desired behavior as I have cached the output of a large vision backbone for all the frames in my dataset, and they are re-fetched whenever necessary.

Regarding timings, I did some deep diving the other day and noticed that indeed, it was because of the network share as you guessed: dataset resides in some other node than the node that the training is conducted, therefore `read` is quite an expensive operation. This was hidden behind some abstractions due to the slurm system and I missed the detail.

And finally, with mmapping - thanks a lot for the tip! I didn't know about the `mmap` package, I will test it and see if there is any performance improvement.

Overall, I will still keep this issue open unless otherwise requested: I do not see the need for looped read operation in the zipfile checks or any other part of the `load` function, particularly for systems where each `read` operation is costly. Maybe those can be combined into a single read, or a way to skip those checks can be added.
